lib/Serialization/Serialization.cpp

```diff
@@ -4390,68 +4263,64 @@ void Serializer::writeAllDeclsAndTypes() {
       writeLocalDeclContext(next);
     }
 
-    while (!DeclContextsToWrite.empty()) {
-      auto next = DeclContextsToWrite.front();
-      DeclContextsToWrite.pop();
+    while (DeclContextsToSerialize.hasMoreToSerialize()) {
+      auto next = DeclContextsToSerialize.popNext(Out.GetCurrentBitNo());
       writeDeclContext(next);
     }
 
-    while (!GenericSignaturesToWrite.empty()) {
-      auto next = GenericSignaturesToWrite.front();
-      GenericSignaturesToWrite.pop();
+    while (GenericSignaturesToSerialize.hasMoreToSerialize()) {
+      auto next = GenericSignaturesToSerialize.popNext(Out.GetCurrentBitNo());
       writeGenericSignature(next);
     }
 
-    while (!GenericEnvironmentsToWrite.empty()) {
-      auto next = GenericEnvironmentsToWrite.front();
-      GenericEnvironmentsToWrite.pop();
-      writeGenericEnvironment(next);
+    while (GenericEnvironmentsToSerialize.hasMoreToSerialize()) {
+      writeNextGenericEnvironment();
     }
 
-    while (!SubstitutionMapsToWrite.empty()) {
-      auto next = SubstitutionMapsToWrite.front();
-      SubstitutionMapsToWrite.pop();
+    while (SubstitutionMapsToSerialize.hasMoreToSerialize()) {
+      auto next = SubstitutionMapsToSerialize.popNext(Out.GetCurrentBitNo());
       writeSubstitutionMap(next);
     }
 
-    while (!NormalConformancesToWrite.empty()) {
-      auto next = NormalConformancesToWrite.front();
-      NormalConformancesToWrite.pop();
+    while (NormalConformancesToSerialize.hasMoreToSerialize()) {
+      auto next = NormalConformancesToSerialize.popNext(Out.GetCurrentBitNo());
       writeNormalConformance(next);
     }
-    
-    while (!SILLayoutsToWrite.empty()) {
-      auto next = SILLayoutsToWrite.front();
-      SILLayoutsToWrite.pop();
+
+    while (SILLayoutsToSerialize.hasMoreToSerialize()) {
+      auto next = SILLayoutsToSerialize.popNext(Out.GetCurrentBitNo());
       writeSILLayout(next);
     }
-  } while (!DeclsAndTypesToWrite.empty() ||
-           !LocalDeclContextsToWrite.empty() ||
-           !DeclContextsToWrite.empty() ||
-           !SILLayoutsToWrite.empty() ||
-           !GenericSignaturesToWrite.empty() ||
-           !GenericEnvironmentsToWrite.empty() ||
-           !SubstitutionMapsToWrite.empty() ||
-           !NormalConformancesToWrite.empty());
+  } while (DeclsToSerialize.hasMoreToSerialize() ||
+           TypesToSerialize.hasMoreToSerialize() ||
+           LocalDeclContextsToSerialize.hasMoreToSerialize() ||
+           DeclContextsToSerialize.hasMoreToSerialize() ||
+           GenericSignaturesToSerialize.hasMoreToSerialize() ||
+           GenericEnvironmentsToSerialize.hasMoreToSerialize() ||
+           SubstitutionMapsToSerialize.hasMoreToSerialize() ||
+           NormalConformancesToSerialize.hasMoreToSerialize() ||
+           SILLayoutsToSerialize.hasMoreToSerialize());
 }
 
-void Serializer::writeAllIdentifiers() {
+std::vector<CharOffset> Serializer::writeAllIdentifiers() {
```

Comment:
IIUC this vector is going to be fairly small so this seems ok... but returning a vector by value seems a bit inconsistent as the typical C++ way would be to take a mutable vector as an argument... 🤷‍♂

Reply:
As of C++11 it'll be moved rather than copied, so it doesn't really make a difference.

Reply:
TIL, thanks! I was under the impression it was a very new feature.